### PR TITLE
DigitalOcean cloud provider and volume plugin

### LINKED ISSUE #32334: support for digital ocean volumes
I was wondering if there are any plans to support digital ocean volumes (similar to awsElasticBlockStore)? If not, are you open to pull requests for this feature?

### PR DESCRIPTION
**Pull request by wardviaene**
**What this PR does / why we need it**: Support for DigitalOcean Cloud and DigitalOcean Volume Plugin

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #32334

**Special notes for your reviewer**: This digitalocean volume plugin is based on the cinder and AWS volume plugin. Some comments with "TODO" have been copied and could be resolved later. I tried to keep the plugin & cloudprovider code as close as possible to other plugins/cloudproviders

**Release note**: Added support for DigitalOcean Cloud and DigitalOcean Volumes

```release-note
```

**Testing**:
You can use the following spec in your pod yml:
```
    spec:
      containers:
      - name: k8s-demo
        image: yourimage
        volumeMounts:
        - mountPath: /myvol
          name: myvolume
        ports:
        - containerPort: 3000
      volumes:
      - name: myvolume
        digitaloceanVolume:
          volumeID: mykubvolume
          fsType: ext4
```

Where mykubvolume is the volume created in DigitalOcean in the same region.

You will need to add create a config file:
```
[Global]
apikey = do-api-key
region = your-region
```

and add these parameters to your kubernetes processes: --cloud-provider=digitalocean --cloud-config=/etc/cloud.config

To test dynamic provisioning, use the following storage class:
```
kind: StorageClass
apiVersion: storage.k8s.io/v1beta1
metadata:
  name: standard
provisioner: kubernetes.io/digitalocean-volume
```

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36894)

### CODE REVIEW

**Comment by wardviaene** [comment]
I signed it!

**Comment by luxas** [comment]
@wlan0 @thockin @roberthbailey 
Please chime in here and point to a design doc or something regarding the new cloud provider mechanism.
If we do not have a full design doc or something like that, we clearly need it. Preferably a proposal.

**Comment by wlan0** [comment]
@wardviaene There is an undergoing effort to remove cloudprovider dependencies out of kubernetes core. Please refer to this proposal for more information -  https://github.com/kubernetes/kubernetes/pull/37037

File: cmd/kube-controller-manager/app/plugins.go [hunk 1]
```diff
@@ -129,6 +132,8 @@ func ProbeControllerVolumePlugins(cloud cloudprovider.Interface, config componen
 			allPlugins = append(allPlugins, azure_dd.ProbeVolumePlugins()...)
 		case photon.ProviderName == cloud.ProviderName():
 			allPlugins = append(allPlugins, photon_pd.ProbeVolumePlugins()...)
+		case digitalocean.ProviderName == cloud.ProviderName():
```
Comment (thockin) [inline cmd/kube-controller-manager/app/plugins.go]: This has gotten out of hand, but we don't have a leg to stand on wrt saying no to this right now.

**Review by thockin — COMMENTED** [review]
I am OK with a volume, but we'd be better off using this to build the pressure on cloud provider extraction.  I don't think we should merge the cloud provider parts.  Can you do those as a standalone controller?

**Comment by wardviaene** [comment]
Ok,  I'll track the refactoring of the cloud API and when finished, I'll redo my PR based on the new proposal

**Comment by luxas** [comment]
@wardviaene So you mean we can close this, right?

**Comment by thockin** [comment]
If you want to resubmit the volume part separately, that's viable...

**Comment by luxas** [comment]
@thockin Why not just wait for Flex Volumes?
I think they should arrive quite soon given the v1.6 goal...

**Comment by thockin** [comment]
I'm more excited about the longer-term flex2 work, but flex1 is/will be
more viable soon

**Comment by klausenbusk** [comment]
Any update on this?

**Comment by thockin** [comment]
The work to extract cloud provider interfaces is progressing.  I am sure
@wlan0 could use help...

**Comment by klausenbusk** [comment]
> The work to extract cloud provider interfaces is progressing. I am sure @wlan0 could use help...

I see, for now I'm trying to solve it with flexvolume (https://github.com/tonyzou/flexvolumes), currently I have done a few changes (support base64 secret, use DigitalOcean go-metadata package to get droplet it, try to detach if already attached) and I'm properly also going to change the mount logic (currently it format the storage every-time it is mounted (not sure why @tonyzou decided that)).
I still have a few issue (detach isn't working properly in the attach function), maybe I should borrow some code from this PR.

**Comment by tonyzou** [comment]
Hi. That looks incredibly dumb and I'm not really sure why I did that either -- I think I was just experimenting. In any case I'd love a pull request.

**Comment by mabounassif** [comment]
@tonyzou @klausenbusk Any updates on that? Any pull requests I can take a look at?

**Comment by klausenbusk** [comment]
> @tonyzou @klausenbusk Any updates on that? Any pull requests I can take a look at?

I have switched to [rexray](https://github.com/codedellemc/rexray/) which has a `flexvolume` plugin. There isn't any 1.6 support [yet](https://github.com/codedellemc/rexray/issues/777), and you could get some [issues]((https://github.com/codedellemc/rexray/issues/795))[1] if you need a lot of volumes (DO only support 5 disks per droplet, so if you need more than 5 disks you could risk that k8s schedule all this pods to the same droplet).

Also you should compile you own version of rexray with [this](https://github.com/codedellemc/libstorage/pull/496) version of libstorage, due to a [bug](https://github.com/digitalocean/godo/pull/136) in DO api.

[1] You can use https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#opaque-integer-resources-alpha-feature as a workaround for now.

**Pull request closed without merging**